skimage/graph/_graph.py

```diff
@@ -70,11 +70,12 @@ def pixel_graph(image, *, mask=None, edge_function=None, connectivity=1, spacing
         The nodes of the graph. These correspond to the raveled indices of the
         nonzero pixels in the mask.
     """
-    if image.dtype == bool and mask is None:
-        mask = image
-    if mask is None and edge_function is None:
-        mask = np.ones_like(image, dtype=bool)
-        edge_function = _weighted_abs_diff
+    if mask is None:
+        if image.dtype == bool:
+            mask = image
+        else:
+            mask = np.ones_like(image, dtype=bool)
+            edge_function = _weighted_abs_diff
```

Comment:
Ah, there's also

https://github.com/scikit-image/scikit-image/blob/fac13fc82b1f2fc9f56c57733e63ea2dd4be97a2/skimage/graph/_graph.py#L121-L122

which also deals with `edge_function` being none because `_weighted_abs_diff` doesn't work with a boolean image. I think we can simplify this confusing logic somewhat. What do you think about 

```patch
Subject: [PATCH] Simplify logic
---
Index: skimage/graph/_graph.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/skimage/graph/_graph.py b/skimage/graph/_graph.py
--- a/skimage/graph/_graph.py	(revision f754e169213df28ca229168c7a8c5325c8a6011d)
+++ b/skimage/graph/_graph.py	(date 1708693357297)
@@ -75,6 +75,12 @@
             mask = image
         else:
             mask = np.ones_like(image, dtype=bool)
+
+    if edge_function is None:
+        if image.dtype == bool:
+            def edge_function(x, y, distances):
+                return distances
+        else:
             edge_function = _weighted_abs_diff
 
     # Strategy: we are going to build the (i, j, data) arrays of a scipy
@@ -119,13 +125,12 @@
     neighbor_indices = neighbors[neighbors_mask]
     neighbor_distances = neighbor_distances_full[neighbors_mask]
     neighbor_indices_sequential = map_array(neighbor_indices, nodes, nodes_sequential)
-    if edge_function is None:
-        data = neighbor_distances
-    else:
-        image_r = image.reshape(-1)
-        data = edge_function(
-            image_r[indices], image_r[neighbor_indices], neighbor_distances
-        )
+
+    image_r = image.reshape(-1)
+    data = edge_function(
+        image_r[indices], image_r[neighbor_indices], neighbor_distances
+    )
+
     m = nodes_sequential.size
     mat = sparse.coo_matrix(
         (data, (indices_sequential, neighbor_indices_sequential)), shape=(m, m)
```

Reply:
Okay, though what about the case when someone tries to use 

```python
skimage.graph.pixel_graph(image, edge_function=my_own_func)
```

With the current implementation, if `image` isn't a boolean, then `my_own_func` is silently replaced by `_weighted_abs_diff`?! Nothing in the docstring suggests this and it seems un-intuitive.

Reply:
True, and it is a bug (unexpected behaviour)! Ok, I can see now that your proposed patch isn't just a refactoring, but a 'full' fix for #7233, in the sense that the current implementation would fix the `TypeError` but not the overall behaviour. So, changing my mind and :+1:

Reply:
@lagru of course! :pray: 
@mao1756 do you think you could expand the current test to ensure, additionally, that `edge_function=func` is *not* silently replaced by `_weighted_abs_diff`? Thank you!